phper/src/types.rs

```diff
@@ -10,162 +10,282 @@
     }
 
     pub const fn is_null(self) -> bool {
-        self.t == IS_NULL
+        self.inner.type_mask == IS_NULL
     }
 
     pub const fn is_bool(self) -> bool {
         self.is_true() || self.is_false()
     }
 
     pub const fn is_true(self) -> bool {
-        get_base_type_by_raw(self.t) == IS_TRUE
+        get_base_type_by_raw(self.inner.type_mask) == IS_TRUE
+    }
+
+    pub const fn is_mixed(self) -> bool {
+        get_base_type_by_raw(self.inner.type_mask) == IS_MIXED
+    }
+
+    pub const fn is_void(self) -> bool {
+        get_base_type_by_raw(self.inner.type_mask) == IS_VOID
+    }
+
+    pub const fn is_never(self) -> bool {
+        get_base_type_by_raw(self.inner.type_mask) == IS_NEVER
+    }
+
+    pub const fn is_callable(self) -> bool {
+        get_base_type_by_raw(self.inner.type_mask) == IS_CALLABLE
+    }
+
+    pub const fn is_iterable(self) -> bool {
+        get_base_type_by_raw(self.inner.type_mask) == IS_ITERABLE
     }
 
     pub const fn is_false(self) -> bool {
-        get_base_type_by_raw(self.t) == IS_FALSE
+        get_base_type_by_raw(self.inner.type_mask) == IS_FALSE
     }
 
     pub const fn is_long(self) -> bool {
-        get_base_type_by_raw(self.t) == IS_LONG
+        get_base_type_by_raw(self.inner.type_mask) == IS_LONG
     }
 
     pub const fn is_double(self) -> bool {
-        get_base_type_by_raw(self.t) == IS_DOUBLE
+        get_base_type_by_raw(self.inner.type_mask) == IS_DOUBLE
     }
 
     pub const fn is_string(self) -> bool {
-        get_base_type_by_raw(self.t) == IS_STRING
+        get_base_type_by_raw(self.inner.type_mask) == IS_STRING
     }
 
     pub const fn is_array(self) -> bool {
-        get_base_type_by_raw(self.t) == IS_ARRAY
+        get_base_type_by_raw(self.inner.type_mask) == IS_ARRAY
     }
 
     pub const fn is_object(self) -> bool {
-        get_base_type_by_raw(self.t) == IS_OBJECT
+        get_base_type_by_raw(self.inner.type_mask) == IS_OBJECT
     }
 
     pub const fn is_resource(self) -> bool {
-        get_base_type_by_raw(self.t) == IS_RESOURCE
+        get_base_type_by_raw(self.inner.type_mask) == IS_RESOURCE
     }
 
     pub const fn is_indirect(self) -> bool {
-        self.t == IS_INDIRECT
+        self.inner.type_mask == IS_INDIRECT
     }
 
     pub const fn get_base_type(self) -> TypeInfo {
-        Self::from_raw(get_base_type_by_raw(self.t))
+        Self::from_type_mask(get_base_type_by_raw(self.inner.type_mask))
     }
 
     pub fn get_base_type_name(self) -> crate::Result<String> {
-        get_type_by_const(self.t)
+        get_type_by_const(self.inner.type_mask)
+    }
+
+    pub fn name(self) -> crate::Result<String> {
+        if !self.has_name() {
+            self.get_base_type_name()
+        } else {
+            // SAFETY: `self.inner.ptr` is not null (or at least shouldn't be...)
+            let z = unsafe { ZStr::from_ptr(self.inner.ptr as *const _zend_string) };
+
+            Ok(z.to_str()?.to_string())
+        }
     }
```

Comment:
not sure about this, `has_name` checks if the `_ZEND_TYPE_NAME_BIT` bit is set, this bit is set if we are creating trhe `TypeInfo` from `ClassEntry` or an object `ZVal`, however, we also expose `from_type_mask` method, which can be used as `let t = TypeInfo::from_type_mask(_ZEND_TYPE_NAME_BIT | IS_STRING)`, meaning we are basically giving the end user a gun to shot their feet with, as this will result in a seg fault if you call `name()` on `t`.

To solve this problem, we have few options:

1. after of checking for `_ZEND_TYPE_NAME_BIT` bit in `TypeInfo::has_name()`, check that `inner.ptr` is not a null pointer
2. make `TypeInfo::from_type_mask` private ( i.e can only be used within a the crate ), and therefor, removing the `impl From<u32> for `TypeInfo`
3. return an `Error::NullPointerError` in `TypeInfo::name()` if `_ZEND_TYPE_NAME_BIT` bit is set, but `ptr` is null.

Reply:
went with option 3